interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+			}
+			switch r2 {
+			case '*', '?', '[', ']', '{', '}', '\\':
+				runes = append(runes, r, r2)
+			default:
+				// do not add r to runes if it's unnecessary
+				runes = append(runes, r2)
+			}
+		case '[', ']', '{', '}':
+			// Should be impossible, we just rendered this variant
+			return nil, fmt.Errorf(`internal error: unexpected unescaped '%v' character`, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	consumeText()
+
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar}) {
+		// If components end with /**/*, strip trailing /*
+		components = components[:len(components)-2]
+	}
+
+	// Add terminal marker or convert existing doublestar to terminal doublestar
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+		components = components[:len(components)-2]
+		components = append(components, component{compType: compSeparatorDoublestarSeparatorTerminal})
+	} else if prevComponentsAre([]componentType{compSeparatorDoublestar}) {
+		components[len(components)-1] = component{compType: compSeparatorDoublestarTerminal}
+	} else {
+		components = append(components, component{compType: compTerminal})
+	}
+
+	var variantBuf strings.Builder
+	var regexBuf strings.Builder
+	regexBuf.WriteRune('^')
+	for _, c := range components {
+		variantBuf.WriteString(c.String())
+		regexBuf.WriteString(c.componentRegex())
+	}
+	regexBuf.WriteRune('$')
+	regex := regexp.MustCompile(regexBuf.String())
+	regex.Longest()
+
+	v := PatternVariant{
+		variant:    variantBuf.String(),
+		components: components,
+		regex:      regex,
+	}
+
+	return &v, nil
+}
+
+// Need to replace unescaped "**", but must be careful about an escaped '\\'
+// before the first '*', since that doesn't escape the first '*'.
+var doublestarReplacer = regexp.MustCompile(`([^\\](\\\\)*)\*\*`)
+
+func prepareVariantForParsing(variant string) string {
+	prepared := doublestarReplacer.ReplaceAllStringFunc(variant, func(s string) string {
+		// Discard trailing "**", add "⁑" instead
+		return s[:len(s)-2] + "⁑"
+	})
+	return prepared
+}
+
+type componentReader struct {
+	components []component
+	submatches []string
+	index      int
+}
+
+func (r *componentReader) next() (*component, string) {
+	if r.index >= len(r.components) {
+		return &component{compType: compTerminal}, ""
+	}
+	comp := &r.components[r.index]
+	submatch := r.submatches[r.index]
+	r.index++
+	return comp, submatch
+}
+
+// Compare returns the relative precence of the receiver and the given pattern
+// variant when considering the given matching path.
+//
+// Returns one of the following, if no error occurs:
+// -1 if v has lower precedence than other
+// 0 if v and other have equal precedence (only possible if v == other)
+// 1 if v has higher precedence than other.
+func (v *PatternVariant) Compare(other *PatternVariant, matchingPath string) (int, error) {
+	selfSubmatches := v.regex.FindStringSubmatch(matchingPath)
+	if selfSubmatches == nil {
```

Comment:
Nitpick, perhaps a switch instead of if-then-else?

```go
switch {
  case selfSubmatches == nil:
     ...
  case len(selfSubmatches)-1 != len(v.components):
     ...
}
```